pkg/target/target_test.go

```diff
@@ -682,3 +694,75 @@ func TestMatcher_Match(t *testing.T) {
 		})
 	}
 }
+
+func TestNamespaceCache(t *testing.T) {
+	ns1 := makeNamespace("my-ns1", map[string]string{"ns1": "label"})
+	ns2 := makeNamespace("my-ns2", map[string]string{"ns2": "label"})
+	tests := []struct {
+		name    string
+		addNs   map[string]interface{}
+		wantNs  map[string]*corev1.Namespace
+		wantErr bool
+	}{
+		{
+			name:    "retrieving a namespace from empty cache returns nil",
+			addNs:   map[string]interface{}{},
+			wantNs:  map[string]*corev1.Namespace{"my-ns2": nil},
+			wantErr: false,
+		},
+		{
+			name:    "retrieving a namespace that doesnt exist returns nil",
+			addNs:   map[string]interface{}{"my-ns1": ns1},
+			wantNs:  map[string]*corev1.Namespace{"my-ns2": nil},
+			wantErr: false,
+		},
+		{
+			name:    "retrieving an added namespace returns the namespace",
+			addNs:   map[string]interface{}{"my-ns1": ns1, "my-ns2": ns2},
+			wantNs:  map[string]*corev1.Namespace{"my-ns1": ns1, "my-ns2": ns2},
+			wantErr: false,
+		},
+		{
+			name:    "adding a non-namespace type returns error",
+			addNs:   map[string]interface{}{"my-ns1": fooConstraint(), "my-ns2": ns2},
+			wantNs:  map[string]*corev1.Namespace{"my-ns2": ns2},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			cache := newNsCache(nil)
```

Comment:
Test this through the `K8sValidationTarget{}` interface. One way would be:

```
cache := newNsCache(nil)
target := &K8sValidationTarget{cache: cache}

...

target.Add(key, ns)
...
target.Remove(key)
```

You're always able to use `cache.Get` later to verify that the state is what we expect.

Reply:
Done. Note: I also changed the API for nsCache.Get and nsCache.Add to accept a namespace instead of an interface{}. This seemed less tedious for the callers, and also to write tests around.